Utilities/ocvalidate/ocvalidate.c

```diff
@@ -34,41 +37,1272 @@
  rm -rf ConfigValidty.dSYM DICT fuzz*.log ConfigValidty
 */
 
+//
+// TODO: Doc functions and add comments where necessary.
+//
 
-long long current_timestamp() {
-    struct timeval te;
-    gettimeofday(&te, NULL); // get current time
-    long long milliseconds = te.tv_sec*1000LL + te.tv_usec/1000; // calculate milliseconds
-    // printf("milliseconds: %lld\n", milliseconds);
-    return milliseconds;
+typedef
+UINT32
+(*CONFIG_CHECK) (
+  IN  OC_GLOBAL_CONFIG  *Config
+  );
+
+STATIC
+INT64
+GetCurrentTimestamp (
+  VOID
+  )
+{
+  struct timeval Time;
+  //
+  // Get current time.
+  //
+  gettimeofday (&Time, NULL);
+  //
+  // Return milliseconds.
+  //
+  return Time.tv_sec * 1000LL + Time.tv_usec / 1000LL;
 }
 
-int main(int argc, char** argv) {
-  uint32_t f;
-  uint8_t *b;
-  if ((b = readFile(argc > 1 ? argv[1] : "config.plist", &f)) == NULL) {
-    printf("Read fail\n");
-    return -1;
+STATIC
+CHAR8 *
+GetFilenameSuffix (
+  IN  CONST CHAR8  *FileName
+  )
+{
+  CONST CHAR8  *SuffixDot;
+
+  //
+  // Find the last dot by reverse order.
+  //
+  SuffixDot = OcAsciiStrrChr (FileName, '.');
+
+  //
+  // In some weird cases the filename can be crazily like
+  // "a.suffix." (note that '.') or "..." (extremely abnormal).
+  // We do not support such.
+  //
+  if (!SuffixDot || SuffixDot == FileName) {
+    return (CHAR8 *) "";
+  }
+
+  //
+  // In most normal cases, return whatever following the dot.
+  //
+  return (CHAR8 *) SuffixDot + 1;
+}
+
+STATIC
+BOOLEAN
+AsciiStringHasAllLegalCharacter (
```

Comment:
Currently this function is called from many places, and does not quite fit the purpose. E.g. path can have spaces, and kernel identifier cannot. Please separate the function in multiple: AsciiCommentHasAllLegalCharacter, AsciiIdentifierHasAllLegalCharacter, AsciiPathHasAllLegalCharacter, etc.

Reply:
Done. https://github.com/acidanthera/OpenCorePkg/pull/166/commits/333b6cbdaefa7c4dc1637d457e96c5d12b450ea5